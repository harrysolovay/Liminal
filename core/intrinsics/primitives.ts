import { assert } from "@std/assert"
import type { Type } from "../Type.ts"
import { declare } from "./declare.ts"

export { null_ as null }
const null_: Type<null> = declare({
  getAtom: () => null_,
  assert: (value) => {
    assert(value === null)
  },
})

export const boolean: Type<boolean> = declare({
  getAtom: () => boolean,
  assert: (value) => {
    assert(typeof value === "boolean")
  },
})

export const integer: Type<number> = declare({
  getAtom: () => integer,
  assert: (value) => {
    assert(Number.isInteger(value))
  },
})

export const number: Type<number> = declare({
  getAtom: () => number,
  assert: (value) => {
    assert(typeof value === "number")
  },
})

export const string: Type<string> = declare({
  getAtom: () => string,
  assert: (value) => {
    assert(typeof value === "string")
  },
})
