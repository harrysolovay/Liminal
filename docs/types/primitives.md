# Primitive Types

```ts twoslash
import { L } from "liminal"
// ---cut---
L.boolean
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.number
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.integer
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.string
// ^?
```

<br />
