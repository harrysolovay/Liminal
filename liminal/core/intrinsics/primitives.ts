import { assert } from "@std/assert"
import { declare } from "../declare.ts"
import type { Type } from "../Type.ts"

export { null_ as null }
const null_: Type<null, never> = declare({
  jsonType: "null",
  getAtom: () => null_,
  assert: (value) => {
    assert(value === null)
  },
})

export const boolean: Type<boolean, never> = declare({
  jsonType: "boolean",
  getAtom: () => boolean,
  assert: (value) => {
    assert(typeof value === "boolean")
  },
})

export const integer: Type<number, never> = declare({
  jsonType: "integer",
  getAtom: () => integer,
  assert: (value) => {
    assert(Number.isInteger(value))
  },
})

export const number: Type<number, never> = declare({
  jsonType: "number",
  getAtom: () => number,
  assert: (value) => {
    assert(typeof value === "number")
  },
})

export const string: Type<string, never> = declare({
  jsonType: "string",
  getAtom: () => string,
  assert: (value) => {
    assert(typeof value === "string")
  },
})
