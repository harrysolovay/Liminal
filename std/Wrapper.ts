import { struct, type Ty } from "../core/mod.ts"

export type Wrapper<X extends Ty = Ty> = ReturnType<typeof Wrapper<X>>

export function Wrapper<X extends Ty>(value: X): struct<{ value: X }> {
  return struct({ value })
}
