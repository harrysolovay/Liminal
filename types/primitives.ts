import { Ty } from "./Ty.ts"

export const none: Ty<null, never> = Ty((description) => ({
  type: "null",
  description,
}))

export const boolean: Ty<boolean, never> = Ty((description) => ({
  type: "boolean",
  description,
}))

export const number: Ty<number, never> = Ty((description) => ({
  type: "number",
  description,
}))

export const string: Ty<string, never> = Ty((description) => ({
  type: "string",
  description,
}))
