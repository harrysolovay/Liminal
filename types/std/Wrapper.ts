import type { Ty } from "../mod.ts"
import { object } from "../object.ts"

export function Wrapper<X extends Ty>(value: X): ReturnType<typeof object<{ value: X }>> {
  return object({ value })
}
