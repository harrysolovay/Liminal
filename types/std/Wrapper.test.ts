import { assertTySnapshot } from "test_util"
import { string } from "../primitives.ts"
import { Wrapper } from "./Wrapper.ts"

Deno.test("Wrapper", async (t) => {
  await assertTySnapshot(t, Wrapper(string))
})
