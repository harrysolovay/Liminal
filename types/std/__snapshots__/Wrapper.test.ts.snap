export const snapshot = {};

snapshot[`Wrapper 1`] = `
{
  additionalProperties: false,
  description: "",
  properties: {
    value: {
      description: "",
      type: "string",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`Wrapper 2`] = `
{
  additionalProperties: false,
  description: "Description one.",
  properties: {
    value: {
      description: "",
      type: "string",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`Wrapper 3`] = `
{
  additionalProperties: false,
  description: "Description two. Description one.",
  properties: {
    value: {
      description: "",
      type: "string",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;
